common/service/sensor/pldm_sensor.c

```diff
@@ -0,0 +1,353 @@
+
+pldm_sensor_thread *pldm_sensor_thread_list;
+pldm_sensor_info *pldm_sensor_list[MAX_SENSOR_THREAD_ID];
+
+__weak pldm_sensor_thread *plat_pldm_sensor_load_thread()
+{
+	return NULL;
+}
+
+__weak pldm_sensor_info *plat_pldm_sensor_load(int thread_id)
+{
+	return NULL;
+}
+
+__weak int plat_pldm_sensor_get_sensor_count(int thread_id)
+{
+	return -1;
+}
+
+bool pldm_sensor_is_interval_ready(pldm_sensor_info *pldm_sensor_list)
+{
+	CHECK_NULL_ARG_WITH_RETURN(pldm_sensor_list, false);
+
+	uint32_t current_time = 0, diff_time = 0;
+
+	current_time = k_uptime_get_32() / 1000;
+	diff_time = current_time - pldm_sensor_list->update_time;
+
+	if (pldm_sensor_list->pdr_numeric_sensor.update_interval > diff_time) {
+		return false;
+	}
+
+	return true;
+}
+
+int pldm_sensor_get_info_via_sensor_id(uint16_t sensor_id, float *resolution, float *offset,
+				       int8_t *unit_modifier, int *cache,
+				       uint8_t *sensor_operational_state)
+{
+	CHECK_NULL_ARG_WITH_RETURN(resolution, -1);
+	CHECK_NULL_ARG_WITH_RETURN(offset, -1);
+	CHECK_NULL_ARG_WITH_RETURN(unit_modifier, -1);
+	CHECK_NULL_ARG_WITH_RETURN(cache, -1);
+	CHECK_NULL_ARG_WITH_RETURN(sensor_operational_state, -1);
+
+	int t_id = 0, s_id = 0, pldm_sensor_count = 0;
+
+	for (t_id = 0; t_id < MAX_SENSOR_THREAD_ID; t_id++) {
+		pldm_sensor_count = plat_pldm_sensor_get_sensor_count(t_id);
+		for (s_id = 0; s_id < pldm_sensor_count; s_id++) {
+			if (sensor_id ==
+			    pldm_sensor_list[t_id][s_id].pdr_numeric_sensor.sensor_id) {
+				// Get from numeric sensor PDR
+				*resolution =
+					pldm_sensor_list[t_id][s_id].pdr_numeric_sensor.resolution;
+				*offset = pldm_sensor_list[t_id][s_id].pdr_numeric_sensor.offset;
+				*unit_modifier = pldm_sensor_list[t_id][s_id]
+							 .pdr_numeric_sensor.unit_modifier;
+				// Get from sensor config
+				*cache = pldm_sensor_list[t_id][s_id].pldm_sensor_cfg.cache;
+				*sensor_operational_state =
+					pldm_sensor_list[t_id][s_id].pldm_sensor_cfg.cache_status;
+				return 0;
+			}
+		}
+	}
+
+	return -1;
+}
+
+uint8_t pldm_sensor_get_reading_from_cache(uint16_t sensor_id, int *reading,
+					   uint8_t *sensor_operational_state)
+{
+	CHECK_NULL_ARG_WITH_RETURN(reading, PLDM_ERROR);
+	CHECK_NULL_ARG_WITH_RETURN(sensor_operational_state, PLDM_ERROR);
+
+	float sensor_reading = 0, decimal = 0, resolution = 0, offset = 0;
+	int cache_reading = 0;
+	int8_t unit_modifier = 0;
+	int16_t integer = 0;
+	uint8_t *sensor_reading_byte;
+
+	if (pldm_sensor_get_info_via_sensor_id(sensor_id, &resolution, &offset, &unit_modifier,
+					       &cache_reading, sensor_operational_state) != 0) {
+		// Couldn't find sensor id in pldm_sensor_list
+		return PLDM_PLATFORM_INVALID_SENSOR_ID;
+	}
+
+	if (resolution == 0) {
+		// The value of resolution couldn't be 0
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	// Convert two byte integer, two byte decimal sensor format to float
+	sensor_reading_byte = malloc(sizeof(int));
+	memcpy(sensor_reading_byte, &cache_reading, sizeof(int));
```

Comment:
Why allocate? Theoritically, you could have gotten away with `uint8_t sensor_reading_byte[sizeof(int)];`
But, assuming we are in little-endian which this is assuming
integer = cache_reading & 0xffff;
decimal = (float)(cache_reading >> 16) / 1000.0;

But I am assuming little endian. I have a feeling this arch is big-endian.

Reply:
If the byte order is off, use a builtin byte swap function: https://docs.zephyrproject.org/latest/doxygen/html/sys_2byteorder_8h.html

Reply:
Done.